cvmfs/catalog_mgr_rw_impl.h

```diff
@@ -0,0 +1,181 @@
+#include "upload.h"
+#include "util.h"
+
+
+
+using namespace std;  // NOLINT
+
+namespace catalog {
+
+template <class CatalogMgrT>
+DirectoryEntryBase
+CatalogBalancer<CatalogMgrT>::CreateEmptyContentDirectoryEntryBase(
+    string name, uid_t uid, gid_t gid) {
+  shash::Algorithms algorithm = catalog_mgr_->spooler_->GetHashAlgorithm();
+  shash::Any file_hash(algorithm);
+  shash::HashString("", &file_hash);
+
+  DirectoryEntryBase deb;
+  deb.name_ = NameString(name);
+  deb.mode_ = S_IFREG | S_IRWXU;
+  deb.size_ = 0;
+  deb.linkcount_ = 1;
+  deb.checksum_ = file_hash;
+  deb.has_xattrs_ = false;
+  deb.mtime_ = time(NULL);
+  deb.inode_ = catalog::DirectoryEntry::kInvalidInode;
+  deb.parent_inode_ = catalog::DirectoryEntry::kInvalidInode;
+  deb.uid_ = uid;
+  deb.gid_ = gid;
+  return deb;
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::AddCvmfsCatalogFile(string path) {
+  XattrList xattr;
+  DirectoryEntry parent;
+  bool retval;
+  retval = catalog_mgr_->LookupPath(PathString(path), kLookupSole, &parent);
+  assert(retval);
+  DirectoryEntryBase cvmfscatalog =
+      CreateEmptyContentDirectoryEntryBase(".cvmfscatalog", parent.uid(),
+                                           parent.gid());
+  DirectoryEntryBase cvmfsautocatalog =
+      CreateEmptyContentDirectoryEntryBase(".cvmfsautocatalog", parent.uid(),
+                                           parent.gid());
+  string relative_path = path.substr(1);
+  catalog_mgr_->AddFile(cvmfscatalog, xattr, relative_path);
+  catalog_mgr_->AddFile(cvmfsautocatalog, xattr, relative_path);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::Balance(catalog_t *catalog) {
+  if (catalog == NULL) {
+    vector<catalog_t*> catalogs = catalog_mgr_->GetCatalogs();
+    for (unsigned i = 0; i < catalogs.size(); ++i)
+      Balance(catalogs[i]);
+    return;
+  }
+  string catalog_path = catalog->path().ToString();
+  virtual_node_t root_node(catalog_path, catalog_mgr_);
+  root_node.ExtractChildren(catalog_mgr_);
+  // we have just recursively loaded the entire virtual tree!
+  OptimalPartition(&root_node);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::OptimalPartition(
+    virtual_node_t *virtual_node) {
+  // postorder track of the fs-tree
+  for (unsigned i = 0; i < virtual_node->children.size(); ++i) {
+    virtual_node_t *virtual_child =
+        &virtual_node->children[i];
+    if (virtual_child->IsDirectory() && !virtual_child->IsCatalog())
+      OptimalPartition(virtual_child);
+  }
+  virtual_node->CaltulateWeight();
+  while (virtual_node->weight > catalog_mgr_->balance_weight_) {
+    typename CatalogBalancer<CatalogMgrT>::VirtualNode *heaviest_node =
+        MaxChild(virtual_node);
+    // we directly add a catalog in this node
+    // TODO(molina) apply heuristics here
+    if (heaviest_node != NULL &&
+        heaviest_node->weight >= catalog_mgr_->min_weight_) {
+      AddCvmfsCatalogFile(heaviest_node->path);
```

Comment:
No, because this process is done in postorder, starting by the leaves, which are the lightest nodes.This way if a father gets overflowed it means that the sum of its children's weight is greater than the balance_weight_ (and not the max_weight_!), so each children must have a weight that is lesser than the balance_weight_, which means it cannot be overflowed.